f3/granite.go

```diff
@@ -350,9 +372,99 @@ func (i *instance) isValid(msg *GMessage) bool {
 	return true
 }
 
-// TODO Checks whether a message is justified by prior messages.
-func (i *instance) isJustified(msg *GMessage) bool {
-	return true
+func (i *instance) VerifyJustification(msg *GMessage) error {
+
+	power := NewStoragePower(0)
+	_ = msg.Justification.QuorumSignature.Signers.ForEach(func(bit uint64) error {
+		power.Add(power, i.powerTable.Entries[bit].Power)
+		return nil
+	})
+
+	if !hasStrongQuorum(power, i.powerTable.Total) {
+		return fmt.Errorf("dropping message as no evidence from a strong quorum: %v", msg.Justification.QuorumSignature.Signers)
+	}
+
+	payload := SignaturePayload(msg.Justification.Payload.Instance, msg.Justification.Payload.Round, msg.Justification.Payload.Step, msg.Justification.Payload.Value)
+	signers := make([]PubKey, 0)
+	_ = msg.Justification.QuorumSignature.Signers.ForEach(func(bit uint64) error {
+		if int(bit) >= len(i.powerTable.Entries) {
+			return nil //TODO handle error
```

Comment:
Please check the error on the outside and error out here. RLE+ has a bit of a sharp edge where it is easy to express "all bits between 0 and 2^64-1" which makes ForEach equivalent to an infinite loop. This is why many methods on it take `max` parameters (if the fact that bit-by-bit iteration happens is unclear). Another way to guard against it is to compare `rle.Count()` against some constant.
We might have to define a constant of `MaxQuorum = 10e6` or something like that (10e6 is arbitrarily large).

Reply:
I changed all `ForEach` to check the bit against the powertable entries and return an error if too big (and propagated the error on the outside). Not sure if anything else is required for the `ForEach` given how we use it.